### PR TITLE
Add MVSECDataLoader with persistent sidecar caching and event backends

### PR DESCRIPTION
**Pull request by toytoi**
Adds the full MVSEC dataloader implementation:

  - `_mvsec_types.py` - Frozen `MVSECOdometryData` dataclass
  - `_mvsec_storage.py` - Content addressed persistent sidecar caches for HDF5 events and GT flow, with two event backends for in memory `CachedEventBackend`, and mmap based `LazyEventBackend`)
  - `_mvsec.py` - `MVSECDataLoader` implementing `DataLoaderBase` for all MVSEC modalities (events, images, IMU, odometry, depth, optical flow, velodyne) with configurable per modality load modes

  This is a large PR, can try to breakup if needed. 
An overview of things to look at:

  1. **`_mvsec_types.py`** (21 lines) - trivial frozen dataclass for odometry 
  2. **`_mvsec_storage.py`** (~700 lines) - self contained caching layer with no public API. Start with `_make_cache_signature` (content addressing scheme), then `_build_event_cache` (atomic temp dir + rename pattern), then the two backend classes at the bottom
  3. **`_mvsec.py`** (~1300 lines) - the main class. `__init__`  are where all modalities are loaded (lines 400-650), then the `DataLoaderBase` interface methods, then rest are the property accessors (mostly trivial one liners)

  Design decisions:

  - Events are stored as typed columns (x/y/t/p) not row tuples, for cache friendly access
  - Sidecar caches use atomic rename (build in temp dir, `os.replace`) so readers never see partial writes
  - HDF5 handles are not safe across `fork()`, so lazy readers track PID and reopen after fork/unpickle
  - Large HDF5 reads are streamed in 1M row blocks to bound peak memory

  ## Tests

100% coverage for all except `_mvsec.py` which has 78%, uncovered lines are optional modality accessors. 151 tests total

---
Next will follow MVSECDataset for PyTorch compatible map/iterable style dataset wrappers that use this dataloader.

### CODE REVIEW

File: src/evlib/dataloaders/_mvsec_storage.py [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+
+ResidentLoadMode = Literal["cached", "lazy"]
+
+# bump these when the on disk sidecar format changes
+_EVENT_CACHE_SCHEMA_VERSION = 1
+_EVENT_BUILD_BLOCK_ROWS = 1_000_000  # keep peak memory bounded during HDF5 reads
+_GT_FLOW_CACHE_SCHEMA_VERSION = 1
+
+
+class _EventCacheMetadata(TypedDict):
```
Comment (shiba24) [inline src/evlib/dataloaders/_mvsec_storage.py]: I feel this EventCache is a bit generic than MVSEC. How about splitting this file into event-cache and flow-cache?

File: src/evlib/dataloaders/_mvsec.py [hunk 2]
```diff
@@ -0,0 +1,1274 @@
...
+    return MVSECOdometryData(
+        timestamps=np.asarray(npz["timestamps"], dtype=np.float64),
+        linear_velocity=np.asarray(npz["lin_vel"], dtype=np.float64),
+        position=np.asarray(npz["pos"], dtype=np.float64),
+        quaternion=np.asarray(npz["quat"], dtype=np.float64),
+        angular_velocity=np.asarray(npz["ang_vel"], dtype=np.float64),
+    )
+
+
+def _find_nearest_index(timestamps: npt.NDArray[np.float64], t: float) -> int:
```
Comment (shiba24) [inline src/evlib/dataloaders/_mvsec.py]: this function could be in `dataloaders/utils.py` (for now in the dataloader folder may be ok, in the future maybe it will be moved out)

File: src/evlib/dataloaders/_mvsec.py [hunk 3]
```diff
@@ -0,0 +1,1274 @@
...
+    sx = np.zeros_like(x_flow, dtype=np.float32)
+    sy = np.zeros_like(y_flow, dtype=np.float32)
+    if np.any(valid):
+        sx[valid] = x_flow[ry[valid], rx[valid]]
+        sy[valid] = y_flow[ry[valid], rx[valid]]
+
+    return sx, sy, valid
+
+
+def _sample_flow_nearest(
```
Comment (shiba24) [inline src/evlib/dataloaders/_mvsec.py]: this could be re-utilized out of the MVSEC dataset, too.

**Review by shiba24 — CHANGES_REQUESTED** [review]
I checked except for the `MVSECDataLoader` class. Could you sort util/generic functions and move them out of MVSEC?

**Comment by toytoi** [comment]
Thanks for the review. 

I went ahead and split `_mvsec_storage.py` into three other modules, 
- `_event_cache.py` has the event sidecar cache logic 
- `_storage_common.py` has shared helpers for things like ResidentLoadMode, _LazyH5Dataset, etc 
- `_mvsec_storage.py` now only has the GT flow cache

For now, I kept the GT flow cache inside `_mvsec_storage.py` as its tied to the MVSEC specific NPZ format. Once DSEC and other datasets are implemented it would be worth considering can be pulled out into a generic flow cache abstraction, but for now I keep it in mvsec. 

I also moved functions like find_nearest_index, sample_flow_nearest, propagate_flow_step, and
  get_flow_coordinate_grid into `dataloaders/utils.py`.

I think in the future, it would be better to move off of the `utils.py` and move into more focused internal modules, like reusable flow helpers and timestamp helpers.

File: src/evlib/dataloaders/_mvsec.py [hunk 4]
```diff
@@ -0,0 +1,1161 @@
...
+from ._storage_common import normalize_resident_load_mode
+from .utils import find_nearest_index
+from .utils import get_flow_coordinate_grid
+from .utils import propagate_flow_step
+
+
+logger = logging.getLogger(__name__)
+
+# False = off, True/"lazy" = on demand, "cached" = in memory
+LoadMode = Union[bool, Literal["lazy", "cached"]]
```
Comment (shiba24) [inline src/evlib/dataloaders/_mvsec.py]: nit: we can use `enum` to define this

File: src/evlib/dataloaders/_mvsec.py [hunk 5]
```diff
@@ -0,0 +1,1161 @@
...
+
+
+# Module lvl helpers
```
Comment (shiba24) [inline src/evlib/dataloaders/_mvsec.py]: nit: `lvl` -> `level`
or just delete this comment + unnecessary empty lines.

File: src/evlib/dataloaders/_mvsec.py (hunk 4 above)
Comment (shiba24) [reply src/evlib/dataloaders/_mvsec.py]:
```
from enum import Enum


class LoadingType(Enum):
    OFF = 0
    LAZY = 1
    CACHED = 2

    @classmethod
    def from_value(cls, value):
        if value is False:
            return cls.OFF
        elif value is True or value == "lazy":
            return cls.LAZY
        elif value == "cached":
            return cls.CACHED
        else:
            raise ValueError(f"Invalid loading type: {value}")
```


then this returns


```
lt = LoadingType.from_value(True)       # LoadingType.LAZY
lt = LoadingType.from_value("lazy")     # LoadingType.LAZY
lt = LoadingType.from_value("cached")   # LoadingType.CACHED
lt = LoadingType.from_value(False)      # LoadingType.OFF
```

File: src/evlib/dataloaders/_mvsec.py [hunk 6]
```diff
@@ -0,0 +1,1161 @@
...
+
+
+def _resolve_load_mode(value: LoadMode) -> Tuple[bool, bool]:
```
Comment (shiba24) [inline src/evlib/dataloaders/_mvsec.py]: please check my comment on enum above

File: src/evlib/dataloaders/_storage_common.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+from typing import cast
+
+import h5py
+import numpy as np
+import numpy.typing as npt
+
+from evlib.codec.fileformat.hdf5 import open_hdf5
+
+
+ResidentLoadMode = Literal["cached", "lazy"]
```
Comment (shiba24) [inline src/evlib/dataloaders/_storage_common.py]: nit: this as enum, too?

**Review by shiba24 — COMMENTED** [review]
Thank you, it's getting closer! I left some minor comments. Could you check them please?

**Comment by toytoi** [comment]
Thanks for the review.

Replaced  `Union[bool, Literal["lazy", "cached"]]` and `_resolve_load_mode()` with `LoadingType` in `_storage_common.py`. I chose to use `str, Enum` so its easier to read in logs and configs rather than using integers. 
                    
Folded `ResidentLoadMode` into `LoadingType` with a `from_resident_value()` classmethod to avoid two overlapping enums.

Removed `# Module lvl helpers` and the unnecessary empty lines. 

`False`, `True`, `"lazy"`, `"cached"` are all still accepted and normalized to `LoadingType`.

**Review by shiba24 — APPROVED** [review]
lgtm. note for the future: i see the potential of file restructuring in the future, util/common parts, and also related tests. e.g.,  currently `TestLoadingType` is in mvsec tests but some can be out of mvsec.

but let's proceed for now. file restructure / refactoring will happen anyways, and we can do this later.

Thank you for the PR!
